### ISSUE TITLE
Unable to reproduce released checkpoint's DynaNav results with Stage-2 training from this repo

### ISSUE THREAD

**Issue by AllenLiu227** [ucla-mobility/TIC-VLA#11]
# Unable to reproduce released checkpoint's DynaNav results with Stage-2 training from this repo

Thanks for releasing the code, dataset and checkpoint — the DynaNav benchmark in particular has been very usable.

We are trying to reproduce the released model and are stuck at roughly **half** its success rate. We think we have isolated the gap to **Stage-2 (action head) training**, and we found some discrepancies between the released checkpoint, the repo configs, and the paper that we cannot resolve from the public materials. We would appreciate clarification.

## Setup

- Training: 4× RTX 6000 Ada, code at current `main`, dataset from `handsomeYun/TIC-VLA` (HF)
- Evaluation: `DynaNav/benchmark.py`, `--navigation_method ticvla`, DynaNav **Hospital, episodes 1–25**, seed 666, Nova Carter, 1.5 m success threshold
- All checkpoints below were evaluated on the **same machine with identical settings** (an RTX 3090; VLM generation latency 5.6–6.8 s, so real `time_delay` is 2–3 s). This depresses absolute numbers for every checkpoint equally, including the released one.

## The core observation

We extracted the VLM weights from the released `TIC-VLA-model.ckpt` (they are present because Stage-2 freezes but still saves the VLM) and trained **our own action head on your Stage-1 VLM**:

| Stage-1 VLM | Stage-2 action head | NE ↓ | SR % ↑ | SPL ↑ | CR % ↓ |
|---|---|---|---|---|---|
| official | **official** (released ckpt) | 4.27 | **80.0** | 0.760 | 16.0 |
| **official** | ours | 11.74 | **40.0** | 0.384 | 40.0 |
| ours | ours | 12.81 | 44.0 | 0.415 | 32.0 |

With **the same Stage-1 VLM**, your Stage-2 produces 80% and ours produces 40%. Swapping in your VLM did not help (40% vs 44% is within noise at n=25).

We verified the VLM extraction is sound:
- 637 VLM tensors load with `missing=0, unexpected=0`
- Evaluated on our own Stage-1 validation split with our code, it scores `val_language_loss = 0.3880`, consistent with the `val_loss=0.3949` in the checkpoint path string
- `vision_model` (340 tensors) is bit-identical to pretrained InternVL3-1B, i.e. correctly frozen

So Stage-1 quality does not appear to be our bottleneck; the difference is in Stage-2 training.

## Discrepancies we found

Read directly from `TIC-VLA-model.ckpt` (`hyper_parameters`, `optimizer_states`, Lightning `loops`):

```
learning_rate:        0.0002          # configs/train_action.yaml says 1.0e-4
weight_decay:         0.01            # from optimizer_states; config says 1.0e-3
action_num_layers:    3               # paper §4 says six cross-attention layers
action_horizon_steps: 30
max_epochs:           10
warmup_steps:         1000
batch_progress/epoch: 3931  == optimizer steps/epoch  → accumulate_grad_batches = 1
val batches/epoch:    207             # 3931/207 = 18.99, consistent with the 5% split
```

**1. Stage-2 batch size / GPU count.** `global_step` gives us one equation with two unknowns:

```
samples/epoch = 3931 × batch_size × num_ranks
```

The repo never states a GPU count (README is a bare `python -m ticvla.training.train`, and `train.py` uses `devices=-1`). With our dataset we get 223,940 training samples, which would require `batch_size × num_ranks ≈ 57` — not a plausible value. Assuming 64 implies ~251k samples; assuming 128 (the paper's 16/GPU × 8 GPUs) implies ~503k. **Could you confirm the Stage-2 batch size and number of GPUs?**

**2. Dataset size.** `vlm_data.py::_find_samples` keeps every 5th frame (hardcoded), which on the released dataset gives:

```
1,177,685 JSON files → 235,726 after stride 5 → 223,940 train / 11,786 val
```

We cannot reconcile 3931 steps/epoch with this file count at any integer stride and any round batch size. **Was the released model trained on the released dataset as-is, and was the stride 5?**

**3. The released checkpoint appears not to have been produced by this repo's code.** The LightningModule constructor signatures differ:

| | Released checkpoint | This repo (`TICVLAActionLightningModule`) |
|---|---|---|
| has | `checkpoint_path`, `device` | `vlm_checkpoint`, `weight_decay` |
| lacks | `weight_decay` | `device` |

The checkpoint's internal paths also point at a `MobileVLA` project (`/data2/yun/MobileVLA/logs/checkpoints_rebuttle/...`). **Is the released Stage-2 training code equivalent to what produced the released checkpoint?** If there are differences (loss masking, target normalisation, optimizer setup), knowing them would help a lot.

**4. `action_num_layers`: 3 vs 6.** The released checkpoint uses 3 (and its 50 `action_expert` tensors match our 3-layer model exactly, shape for shape). The paper says six cross-attention layers. We assume the checkpoint is authoritative, but wanted to flag the inconsistency.

## Things we ruled out on our side

In case it saves anyone else time, these were checked and are *not* the cause:

- **Action-head architecture** — our `action_expert` is tensor-for-tensor identical to the released one (50 tensors, all shapes match; trained weight std within ~1.2× median)
- **KV-cache dims** — the hardcoded `kv_num_heads=2, kv_head_dim=64 → 128` matches InternVL3-1B's `num_key_value_heads=2`, `hidden_size/num_attention_heads = 896/14 = 64`
- **`time_delay` distribution** — training samples `U(0, min(10, elapsed))`, which covers our 2–3 s inference latency
- **Waypoint coordinate frame** — training targets and inference outputs are both current-frame-relative `(dx, dy)`
- **Zero-padded action targets** — `_extract_waypoints_xy` zero-fills when `len(future) < 30` and the `smooth_l1_loss` is unmasked, so those steps train the model toward "don't move". Real, but it affects only **0.8% of samples / 0.4% of target steps**, too rare to explain the gap.

We also observed that `val_total_loss` is not usable for selecting across training runs: our best run reaches `val_ade = 0.135` versus the released checkpoint's `0.189`, while scoring 40% against its 80%. (Within a single run it does track closed-loop performance.) This may be worth a note in the README for others reproducing the work.

## What would help most

1. Stage-2 `batch_size` and GPU count (to resolve the 3931 steps/epoch)
2. Whether the released dataset and stride match what trained the released model
3. Whether the released Stage-2 code matches the code that produced the released checkpoint, and if not, what differs

Happy to share full logs, per-episode benchmark JSON, or run additional diagnostics if useful. Thanks!
